src/common/XAuth.cpp

```diff
@@ -95,33 +102,65 @@ bool XAuth::addCookie(const QString &display)
         return false;
     }
 
-    return XAuth::addCookieToFile(display, m_authPath, m_cookie);
+    return XAuth::writeCookieToFile(display, authPath(), m_cookie);
 }
 
-bool XAuth::addCookieToFile(const QString &display, const QString &fileName,
-                            const QString &cookie)
+bool XAuth::writeCookieToFile(const QString &display, const QString &fileName,
+                              QByteArray cookie)
 {
-    qDebug() << "Adding cookie to" << fileName;
 
-    // Touch file
-    QFile file_handler(fileName);
-    file_handler.open(QIODevice::Append);
-    file_handler.close();
+    qDebug() << "Writing cookie to" << fileName;
+
+    if(display.size() < 2 || display[0] != QLatin1Char(':') || cookie.size() != 16)
+        return false;
+
+    // The file needs 0600 permissions
+    int oldumask = umask(077);
+
+    // Truncate the file. We don't support merging like the xauth tool does.
+    FILE * const authFp = fopen(qPrintable(fileName), "wb");
+    umask(oldumask);
+    if (authFp == nullptr)
+        return false;
+
+    char localhost[HOST_NAME_MAX + 1] = "";
+    if (gethostname(localhost, sizeof(localhost)) < 0)
+        strcpy(localhost, "localhost");
+
+    ::Xauth auth = {};
+    char cookieName[] = "MIT-MAGIC-COOKIE-1";
 
-    QString cmd = QStringLiteral("%1 -f %2 -q").arg(mainConfig.X11.XauthPath.get()).arg(fileName);
+    // Skip the ':'
+    QByteArray displayNumberUtf8 = display.midRef(1).toUtf8();
 
-    // Execute xauth
-    FILE *fp = ::popen(qPrintable(cmd), "w");
+    auth.family = FamilyLocal;
+    auth.address = localhost;
+    auth.address_length = strlen(auth.address);
+    auth.number = displayNumberUtf8.data();
+    auth.number_length = displayNumberUtf8.size();
+    auth.name = cookieName;
+    auth.name_length = sizeof(cookieName) - 1;
+    auth.data = cookie.data();
+    auth.data_length = cookie.size();
 
-    // Check file
-    if (!fp)
+    if (XauWriteAuth(authFp, &auth) == 0) {
+        fclose(authFp);
```

Comment:
I would suggest using QScopeGuard rather than having to fClose in all the paths

Reply:
Done